RavenDB-15919 - Concurrent subscriptions

lastav5 opened the pull request:
https://issues.hibernatingrhinos.com/issue/RavenDB-15919

ayende commented on src/Raven.Client/Documents/Subscriptions/SubscriptionWorker.cs:
```diff
@@ -512,6 +515,13 @@ private async Task ProcessSubscriptionAsync()
                             {
                                 if (tcpStreamCopy != null) //possibly prevent ObjectDisposedException
                                 {
+                                    if (_forTestingPurposes?.CloseThisWorkerBeforeAck == 2)
```
what is `2` ?

ayende commented on src/Raven.Client/Documents/Subscriptions/SubscriptionWorker.cs:
```diff
@@ -877,6 +888,8 @@ internal TestingStuff ForTestingPurposesOnly()
         {
             internal bool SimulateUnexpectedException;
+            internal int CloseThisWorkerBeforeAck;
+            internal ManualResetEventSlim mre = new ();
```
naming convention

ayende commented on src/Raven.Server/Documents/Handlers/SubscriptionsHandler.cs:
```diff
@@ -228,16 +228,18 @@ await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBody
+                    return;
+                }
+
+                //TODO: return all connections' details
```
What remains to be done here? Also, what about backward compatability?

ppekrol replied:
this is not exposed in the Client API, at this point we should definitely fix this enpoint to our default format:

```
{
    "Results": [ {}, {} ]
}

```

ayende commented on src/Raven.Server/Documents/Subscriptions/ConcurrentSubscription.cs:
```diff
@@ -0,0 +1,506 @@
+            }
+            else
+            {
+                IsConcurrent = false;
```
So you can have `ConcurrentSubscription.IsConcurrent == false` ? The naming should be improved.
Isn't this the same as `SubscriptionConnectionState` previously?

ayende commented on src/Raven.Server/Documents/Subscriptions/ConcurrentSubscription.cs:
```diff
@@ -0,0 +1,506 @@
+            _subscriptionStorage = storage;
+
+            CancellationTokenSource = new();
+            _waitForMoreDocuments = new AsyncManualResetEvent(CancellationTokenSource.Token);//TODO stav: check what is being done with this token
```
You probably need to link the token to at least the server shutdown  / database shutdown tokens

ayende commented on src/Raven.Server/Documents/Subscriptions/ConcurrentSubscription.cs:
```diff
@@ -0,0 +1,506 @@
+
+            _docsLock = new SemaphoreSlim(1);
+            LastChangeVectorSent = connection.SubscriptionState.ChangeVectorForNextBatchStartingPoint;
+            //SetLastChangeVectorSent(connection.SubscriptionState.ChangeVectorForNextBatchStartingPoint); //TODO stav: Should use this even though there is no lock here?
```
There is a lock, on the list of overall connections, no?

ayende commented on src/Raven.Server/Documents/Subscriptions/ConcurrentSubscription.cs:
```diff
@@ -0,0 +1,506 @@
+            LastChangeVectorSent = connection.SubscriptionState.ChangeVectorForNextBatchStartingPoint;
+            //SetLastChangeVectorSent(connection.SubscriptionState.ChangeVectorForNextBatchStartingPoint); //TODO stav: Should use this even though there is no lock here?
+
+            //TODO stav: make sure to handle a scenario when this object gets created but the first(this) connection fails to be registered
```
how can that happen?

ayende commented on src/Raven.Server/Documents/Subscriptions/ConcurrentSubscription.cs:
```diff
@@ -0,0 +1,506 @@
+        public IEnumerable<(Document Doc, Exception Exception)> GetNextBatch(SubscriptionConnection connection, SubscriptionDocumentsFetcher fetcher,
+            DocumentsOperationContext docsContext, IncludeDocumentsCommand includesCmd)
+        {
+            //TODO stav: create fetcher here instead of passing it as an argument
```
why?

ayende commented on src/Raven.Server/Documents/Subscriptions/ConcurrentSubscription.cs:
```diff
@@ -0,0 +1,506 @@
+        private readonly SubscriptionStorage _subscriptionStorage;
+        private readonly DocumentsStorage _documentsStorage;
+        private readonly SubscriptionState _subscriptionState;
+        private LimitedConcurrentDictionary<SubscriptionConnection> _connections;
```
why is this limited size one?

ayende commented on src/Raven.Server/Documents/Subscriptions/ConcurrentSubscription.cs:
```diff
@@ -0,0 +1,506 @@
+                {
+                    var resendDocs = clusterStateMachine.GetSubscriptionDocumentsThatWerePreviouslySentAndNotInAnyActiveBatches(clusterOperationContext,
+                        connection.TcpConnection.DocumentDatabase.Name, SubscriptionId, _connections.Select(conn => conn.CurrentBatchId).
+                            Where(id => id != 0).ToHashSet()); //TODO stav: need to make sure batchId-to-ignore is a consistent value globally
```
Let's avoid Linq here.
Also, instead of `HashSet`, can pass a predicate to avoid the allocation

ayende commented on src/Raven.Server/Documents/Subscriptions/ConcurrentSubscription.cs:
```diff
@@ -0,0 +1,506 @@
+                            yield break;
+                        }
+
+                        //TODO stav: ActiveBatches may change externally - make concurrent
```
That is actually a good point. We need a consistent state of the `ActiveBatches`, might still make sense to create a `HashSet` in that case. But since Subscription is long lived, let's allocate it once.

ayende commented on src/Raven.Server/Documents/Subscriptions/ConcurrentSubscription.cs:
```diff
@@ -0,0 +1,506 @@
+                        //TODO stav: ActiveBatches may change externally - make concurrent
+                        if (clusterStateMachine.IsDocumentInActiveBatch(clusterOperationContext, connection.TcpConnection.DocumentDatabase.Name, SubscriptionId, doc.Id,
+                            _connections.Select(conn => conn.CurrentBatchId).Where(id => id!=0)//TODO stav: need to make sure batchId to ignore is a consistent value globally
+                                .ToHashSet()))
```
avoiding linq

ayende commented on src/Raven.Server/Documents/Subscriptions/ConcurrentSubscription.cs:
```diff
@@ -0,0 +1,506 @@
+        public void WaitForDocsLock()
+        {
+            Console.WriteLine("wait for lock");
+            _docsLock.Wait(); //TODO stav: currently locks the thread
```
What is this doing? Shouldn't this be async?

ayende commented on src/Raven.Server/Documents/Subscriptions/ConcurrentSubscription.cs:
```diff
@@ -0,0 +1,506 @@
+                Console.WriteLine($"Trying to register Connection {incomingConnection.ClientUri} with strategy: {incomingConnection.Strategy}");
+                if (IsConcurrent)
+                {
+                    switch (incomingConnection.Strategy)
```
would be simpler checking if this `Concurrent`, no?

ayende commented on src/Raven.Server/Documents/Subscriptions/ConcurrentSubscription.cs:
```diff
@@ -0,0 +1,506 @@
+                                    CancelSingleConnection(currentConnection, new SubscriptionInUseException("Closed by TakeOver"));
+                                    return GetDisposingAction();
+                                }
+                                //TODO stav: there is no connection connected currently. throw here or takeover anyway?
```
If this is null, can just add, no?

ayende commented on src/Raven.Server/Documents/Subscriptions/ConcurrentSubscription.cs:
```diff
@@ -0,0 +1,506 @@
+                    {
+                        Console.WriteLine($"Delayed {incomingConnection.ClientUri}");
+                        await Task.Delay(timeToWait);
+                        throw new TimeoutException();
```
I know that this is from the old code, but it makes no sense whatsoever

ayende commented on src/Raven.Server/Documents/Subscriptions/ConcurrentSubscription.cs:
```diff
@@ -0,0 +1,506 @@
+                                if (currentConnection != null)
+                                {
+                                    TryRemoveConnection(currentConnection);
+                                    TryAddConnection(incomingConnection);
```
That isn't under any lock, is it?

ayende commented on src/Raven.Server/Documents/Subscriptions/ConcurrentSubscription.cs:
```diff
@@ -0,0 +1,506 @@
+            {
+                return new DisposeOnce<SingleAttempt>(() =>
+                {
+                    while (_recentConnections.Count > 10)
```
what if we have 10 concurrent subscriptions?

ayende commented on src/Raven.Server/Documents/Subscriptions/ConcurrentSubscription.cs:
```diff
@@ -0,0 +1,506 @@
+        }
+
+        //TODO stav: when should this be called? check original code
+        public void CleanupRecentAndRejectedConnections()
```
this is never called, can remove

lastav5 replied:
Done

ayende commented on src/Raven.Server/Documents/Subscriptions/ConcurrentSubscription.cs:
```diff
@@ -0,0 +1,506 @@
+
+        public void DropSingleConnection(SubscriptionConnection connection)
+        {
+            if (_connections.Contains(connection))
```
why not just `TryRemove` ?

ayende commented on src/Raven.Server/Documents/Subscriptions/ConcurrentSubscription.cs:
```diff
@@ -0,0 +1,506 @@
+                        {
+                            // ignored
+                        }
+                    }
```
Probably not, there shouldn't be a scenario where we throw here.

lastav5 replied:
Done

ayende commented on src/Raven.Server/Documents/Subscriptions/ConcurrentSubscription.cs:
```diff
@@ -0,0 +1,506 @@
+
+        public SubscriptionConnection MostRecentEndedConnection()
+        {
+            if (_recentConnections.TryPeek(out var recentConnection))
```
this is the _oldest_ connection

ayende commented on src/Raven.Server/Documents/Subscriptions/ConcurrentSubscription.cs:
```diff
@@ -0,0 +1,506 @@
+                    // ignored: If we've failed to raise the cancellation token, it means that it's already raised
+                }
+            }
+            //TODO stav: also cancel this.CancellationTokenSource?
```
Yes, let's abort everything.

ayende commented on src/Raven.Server/Documents/Subscriptions/SubscriptionStorage.cs:
```diff
@@ -268,30 +289,34 @@ public async Task DeleteSubscription(string name, string raftRequestId)
-                {
-                    // ignored
-                }
+                concurrentSubscription.CancelSingleConnection(subscriptionConnection, ex); //TODO stav: Could cause problems concurrently?
```
Under what scenario this is called?
Should just dispose `concurrentSubscription`, no? However, user may want to close a _single_ connection, how do we do it?

ayende commented on src/Raven.Server/Documents/Subscriptions/SubscriptionStorage.cs:
```diff
@@ -268,30 +289,34 @@ public async Task DeleteSubscription(string name, string raftRequestId)
+            if (_subscriptions.TryGetValue(subscriptionId, out ConcurrentSubscription concurrentSubscription) == false)
+                return false;
+
+            if(concurrentSubscription.GetConnections().Contains(connection))//TODO stav: Could cause problems concurrently?
```
Yes, don't do things like that, instead, `TryCancel`

ayende commented on src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs:
```diff
@@ -140,20 +146,21 @@ public void AddToStatusDescription(string message)
             CancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(TcpConnection.DocumentDatabase.DatabaseShutdown);
             _supportedFeatures = TcpConnectionHeaderMessage.GetSupportedFeaturesFor(TcpConnectionHeaderMessage.OperationTypes.Subscription, connectionOptions.ProtocolVersion);
-            _waitForMoreDocuments = new AsyncManualResetEvent(CancellationTokenSource.Token);
+            //_waitForMoreDocuments = new AsyncManualResetEvent(CancellationTokenSource.Token);
```
?

ayende commented on src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs:
```diff
@@ -247,21 +256,21 @@ private async Task InitAsync()
                             _logger.Info(
                                 $"A connection from IP {TcpConnection.TcpClient.Client.RemoteEndPoint} is starting to wait until previous connection from " +
-                                $"{_connectionState.Connection?.TcpConnection.TcpClient.Client.RemoteEndPoint} is released");
+                                $"{_subscriptionConnections.GetConnections().First()?.TcpConnection.TcpClient.Client.RemoteEndPoint} is released");
```
This is wrong

ayende commented on src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs:
```diff
@@ -247,21 +256,21 @@ private async Task InitAsync()
                         await SendHeartBeat(
                             $"A connection from IP {TcpConnection.TcpClient.Client.RemoteEndPoint} is waiting for Subscription Task that is serving a connection from IP " +
-                            $"{_connectionState.Connection?.TcpConnection.TcpClient.Client.RemoteEndPoint} to be released");
+                            $"{_subscriptionConnections.GetConnections().First()?.TcpConnection.TcpClient.Client.RemoteEndPoint} to be released");
```
this is wrong

ayende commented on src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs:
```diff
@@ -315,88 +324,92 @@ private async Task FlushBufferToNetwork()
+                        if (gotSemaphore)
+                        {
+                            //TODO stav: release semaphore when?
+                            //TcpConnection.DocumentDatabase.SubscriptionStorage.ReleaseSubscriptionsSemaphore();
```
This is here to limit the number of concurrent subscriptions.
You release this if you got it, as previously state. Note that the default is `1000` connections (across *all* subscriptions) on the database node

ayende commented on src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs:
```diff
@@ -669,26 +645,27 @@ private async Task ProcessSubscriptionAsync()
                                         // if this is a new subscription that we sent anything in this iteration,
                                         // _lastChangeVector is null, so let's not change it
-                                        _lastChangeVector ??
+                                        _subscriptionConnections.LastChangeVectorSent ?? //TODO stav: send only DoNotChange?
```
I don't understand the question

ayende commented on src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs:
```diff
@@ -669,26 +645,27 @@ private async Task ProcessSubscriptionAsync()
                                         nameof(Client.Constants.Documents.SubscriptionChangeVectorSpecialStates.DoNotChange),
-                                        subscriptionChangeVectorBeforeCurrentBatch);
+                                        subscriptionChangeVectorBeforeCurrentBatch,
+                                        0, new HashSet<string>());//TODO stav: setup something so table in cluster won't change
```
why?

ayende commented on src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs:
```diff
@@ -829,178 +840,219 @@ private void AssertCloseWhenNoDocsLeft()
-            bool anyDocumentsSentInCurrentIteration = false;
-
-            using (batchScope.For(SubscriptionOperationScope.BatchSendDocuments))
+            if (await _subscriptionConnections.WaitForDocsLock(300) == false)
```
Where did the `300` came from? What is that about?

ayende commented on src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs:
```diff
@@ -829,178 +840,219 @@ private void AssertCloseWhenNoDocsLeft()
+                                //_startEtag = result.Doc.Etag;
+                                //lastChangeVectorSentInThisBatch = string.IsNullOrEmpty(SubscriptionState.ChangeVectorForNextBatchStartingPoint)
+                                //    ? result.Doc.ChangeVector
+                                //    : ChangeVectorUtils.MergeVectors(result.Doc.ChangeVector, SubscriptionState.ChangeVectorForNextBatchStartingPoint);
```
why?

ayende commented on src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs:
```diff
@@ -829,178 +840,219 @@ private void AssertCloseWhenNoDocsLeft()
+                                    var batchId = await TcpConnection.DocumentDatabase.SubscriptionStorage.RecordBatchDocuments(
+                                        _subscriptionConnections.SubscriptionId,
+                                        _currentBatchDocumentIdToChangeVector);
+                                    CurrentBatchId = batchId;
```
The documents were already sent at this point? Can another subscription operate here now?

ayende commented on src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs:
```diff
@@ -1059,25 +1111,36 @@ private async Task FlushDocsToClient(AsyncBlittableJsonTextWriter writer, int fl
             {
-                var hasMoreDocsTask = _waitForMoreDocuments.WaitAsync();
-                var resultingTask = await Task.WhenAny(hasMoreDocsTask, pendingReply, TimeoutManager.WaitFor(TimeSpan.FromMilliseconds(WaitForChangedDocumentsTimeoutInMs))).ConfigureAwait(false);
+                var hasMoreDocsTask = _subscriptionConnections.WaitForMoreDocs();
```
shouldn't this be `await`ed?

ayende commented on src/Raven.Server/ServerWide/ClusterStateMachine.cs:
```diff
@@ -311,6 +343,11 @@ protected override void Apply(ClusterOperationContext context, BlittableJsonRead
+                    //case nameof(RecordBatchSubscriptionDocumentsCommand):
+                    //    ExecuteRecordBatchSubscriptionDocuments(context, cmd, index, type);
+                    //    break;
+
```
??

ayende commented on src/Raven.Server/ServerWide/ClusterStateMachine.cs:
```diff
@@ -662,6 +703,196 @@ void AddUnrecoverableNotification(string error, Exception exception)
+            }
+
+            var subscriptionState = context.Transaction.InnerTransaction.OpenTable(SubscriptionStateSchema, SubscriptionState);
+            //Voron.Debugging.DebugStuff.RenderAndShow(subscriptionState.GetTree(SubscriptionStateKeySlice, true));
```
remove

ayende commented on src/Raven.Server/ServerWide/ClusterStateMachine.cs:
```diff
@@ -662,6 +703,196 @@ void AddUnrecoverableNotification(string error, Exception exception)
+            }
+        }
+
+        //private void ExecuteRecordBatchSubscriptionDocuments(ClusterOperationContext context, BlittableJsonReaderObject cmd, long index, string type)
```
??

ayende commented on src/Raven.Server/ServerWide/Commands/Subscriptions/AcknowledgeSubscriptionBatchCommand.cs:
```diff
@@ -59,14 +64,20 @@ protected override BlittableJsonReaderObject GetUpdatedValue(long index, RawData
-
-            subscription.ChangeVectorForNextBatchStartingPoint = ChangeVector;
+            //TODO stav: handle this check?
+            //if (LastKnownSubscriptionChangeVector != subscription.ChangeVectorForNextBatchStartingPoint)
```
No, because we now have concurrency here. Need to think what the scenario here should be.

ayende commented on src/Sparrow.Server/Collections/LimitedConcurrentDictionary.cs:
```diff
@@ -0,0 +1,75 @@
+
+namespace Sparrow.Server.Collections
+{
+    public class LimitedConcurrentDictionary<T> : ConcurrentSet<T>
```
This looks like a semaphore, but I don't understand what you are trying to do here.

ayende commented on src/Sparrow/Collections/ConcurrentSet.cs:
```diff
@@ -42,7 +43,7 @@ public void Add(T item)
         }
 
-        public bool TryAdd(T item)
+        public virtual bool TryAdd(T item)
```
let's avoid adding virtual method call overhead here.

ayende commented on src/Sparrow/Collections/ConcurrentSet.cs:
```diff
@@ -52,7 +53,7 @@ public bool Contains(T item)
         }
 
-        public bool TryRemove(T item)
+        public virtual bool TryRemove(T item)
```
let's avoid adding virtual method call overhead here.

ppekrol commented on src/Raven.Client/Documents/Subscriptions/SubscriptionWorker.cs:
```diff
@@ -19,6 +19,7 @@
 using System.IO;
 using System.Linq;
 using System.Net.Sockets;
+//using System.Reflection.Metadata.Ecma335;
```
remove?

ppekrol commented on src/Raven.Client/Documents/Subscriptions/SubscriptionWorker.cs:
```diff
@@ -216,7 +217,9 @@ await requestExecutor.ExecuteAsync(_redirectNode, null, context, command, should
                 }
-                
+
+                Console.WriteLine($"Connect to {tcpInfo.NodeTag}");
```
remove?

ppekrol commented on test/SlowTests/Client/Subscriptions/RavenDB_3491.cs:
```diff
@@ -285,7 +285,7 @@ public async Task SubscribtionWithEtag_MultipleOpens()
                 }
 
-                Assert.True(Task.WaitAll(new[] {subscriptionReleasedAwaiter}, 250));
+                //Assert.True(Task.WaitAll(new[] {subscriptionReleasedAwaiter}, 250));
```
hm?

lastav5 replied:
fixed

ppekrol commented on src/Raven.Server/Documents/Subscriptions/SubscriptionStorage.cs:
```diff
@@ -412,49 +438,49 @@ public SubscriptionGeneralDataAndStats GetRunningSubscription(TransactionOperati
         {
-            public SubscriptionConnection Connection;
+            public List<SubscriptionConnection> Connections;
+            public SubscriptionConnection Connection => Connections?.FirstOrDefault();
```
probably can be removed but requires Studio change, not used anywhere else

ayende replied:
Can we make the studio change, then?

ppekrol commented on src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs:
```diff
@@ -247,21 +254,22 @@ private async Task InitAsync()
                             _logger.Info(
                                 $"A connection from IP {TcpConnection.TcpClient.Client.RemoteEndPoint} is starting to wait until previous connection from " +
-                                $"{_connectionState.Connection?.TcpConnection.TcpClient.Client.RemoteEndPoint} is released");
+                                $"{_subscriptionConnectionsState.GetConnections().First()?.TcpConnection.TcpClient.Client.RemoteEndPoint} is released");
```
Is this is an issue? need to take into account more connections?

ayende replied:
In theory, to get here you can only have a single previous connection, but lets change the code to report on all of them, just to be clear

ppekrol commented on src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs:
```diff
@@ -247,21 +254,22 @@ private async Task InitAsync()
                         await SendHeartBeat(
                             $"A connection from IP {TcpConnection.TcpClient.Client.RemoteEndPoint} is waiting for Subscription Task that is serving a connection from IP " +
-                            $"{_connectionState.Connection?.TcpConnection.TcpClient.Client.RemoteEndPoint} to be released");
+                            $"{_subscriptionConnectionsState.GetConnections().First()?.TcpConnection.TcpClient.Client.RemoteEndPoint} to be released");
```
Is this is an issue? need to take into account more connections?

ppekrol commented on src/Raven.Server/Documents/Subscriptions/SubscriptionStorage.cs:
```diff
@@ -578,28 +598,28 @@ internal void CleanupSubscriptions()
+            foreach (var state in _subscriptions)
             {
-                if (state.Value.Connection != null)
+                if (state.Value.GetConnections().Count != 0)
```
use dedicated method

ppekrol commented on src/Raven.Server/Documents/Subscriptions/SubscriptionStorage.cs:
```diff
@@ -578,28 +598,28 @@ internal void CleanupSubscriptions()
+            foreach (var kvp in _subscriptions)
             {
-                if (state.Value.Connection != null)
+                if (kvp.Value.GetConnections().Count != 0) //TODO: what if connections is empty only for a second (between switches) - make a IsEmpty func with timeout?
```
use dedicated method

ppekrol commented on src/Raven.Server/Documents/Subscriptions/SubscriptionStorage.cs:
```diff
@@ -518,51 +544,45 @@ public void HandleDatabaseRecordChange(DatabaseRecord databaseRecord)
-                    if (connection != null && subscriptionState.Query != connection.SubscriptionState.Query)
+                    //make sure we only drop old connection and not new ones just arriving with the updated query
+                    SubscriptionConnectionsState subscriptionConnectionsState = subscriptionStateKvp.Value;
+                    if (subscriptionConnectionsState != null && subscriptionState.Query != subscriptionConnectionsState.Query)
```
double-check if this is OK?

ayende replied:
Looks ok

ppekrol commented on src/Raven.Server/Documents/Subscriptions/SubscriptionStorage.cs:
```diff
@@ -476,7 +502,7 @@ public SubscriptionGeneralDataAndStats(SubscriptionState @base)
         public long GetRunningCount()
         {
-            return _subscriptionConnectionStates.Count(x => x.Value.Connection != null);
+            return _subscriptions.Count(x => x.Value.GetConnections() != null && x.Value.GetConnections().Count != 0);//TODO stav: indicator for running?
```
use dedicated method

ppekrol commented on src/Raven.Server/Documents/Subscriptions/SubscriptionStorage.cs:
```diff
@@ -268,30 +291,35 @@ public async Task DeleteSubscription(string name, string raftRequestId)
+        }
+
+        //TODO stav: can remove
+        public bool DropSubscriptionConnection(long subscriptionId, SubscriptionConnection connection, SubscriptionException ex)
```
remove

ppekrol commented on src/Raven.Server/Documents/Subscriptions/SubscriptionStorage.cs:
```diff
@@ -268,30 +291,35 @@ public async Task DeleteSubscription(string name, string raftRequestId)
+        public bool DropSubscriptionConnections(long subscriptionId, SubscriptionException ex)
         {
-            if (_subscriptionConnectionStates.TryGetValue(subscriptionId, out SubscriptionConnectionState subscriptionConnectionState) == false)
+            if (_subscriptions.TryRemove(subscriptionId, out SubscriptionConnectionsState concurrentSubscription) == false)
```
should we dispose the instance maybe?

ppekrol commented on src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionsState.cs:
```diff
@@ -0,0 +1,500 @@
+            try
+            {
+                EndConnections();
+                CancellationTokenSource.Cancel();
```
change the order, token needs to be canceled first

ppekrol commented on src/Raven.Server/ServerWide/ClusterCommandsVersionManager.cs:
```diff
@@ -68,6 +68,7 @@ public static class ClusterCommandsVersionManager
             [nameof(UpdateExternalReplicationStateCommand)] = Base40CommandsVersion,
             [nameof(UpdateTopologyCommand)] = Base40CommandsVersion,
             [nameof(AcknowledgeSubscriptionBatchCommand)] = Base40CommandsVersion,
+            [nameof(RecordBatchSubscriptionDocumentsCommand)] = Base40CommandsVersion,
```
should be 53_000

lastav5 committed:
RavenDB-15919 - Concurrent Subscriptions. WIP

karmeli87 committed:
RavenDB-15919 fixes

lastav5 committed:
RavenDB-15919 - Code cleanup and fixes. Adressed PR comments

karmeli87 committed:
RavenDB-15919 abstract the document fetcher

karmeli87 committed:
RavenDB-15919 fixes

lastav5 committed:
RavenDB-15919 - Fixed subscription details endpoint

karmeli87 committed:
RavenDB-15919 fixes after refactoring

also fix revisions include with JS

lastav5 committed:
RavenDB-15919 - Fixed subscription-details endpoint when no connections exist. Added endpoint support for dropping a single connection by Id

karmeli87 commented on src/Raven.Server/Documents/Handlers/SubscriptionsHandler.cs:
```diff
@@ -386,18 +399,30 @@ public Task DropSubscriptionConnection()
         {
             var subscriptionId = GetLongQueryString("id", required: false);
             var subscriptionName = GetStringQueryString("name", required: false);
+            var connectionUri = GetStringQueryString("connectionUri", required: false);
```
let's have some UID per connection and pass it here

lastav5 committed:
RavenDB-15919 - fixed subscription connection drop and updated client API

ppekrol commented on src/Raven.Client/Documents/Subscriptions/DocumentSubscriptions.cs:
```diff
@@ -481,17 +481,43 @@ public List<SubscriptionState> GetSubscriptions(int start, int take, string data
         /// <param name="id"></param>
         /// <param name="database"></param>
-        public void DropConnection(string name, string database = null)
+        public void DropConnection(string name, long connectionId, string database = null)
```
breaks source-level compatibility

lastav5 replied:
Fixed

ppekrol commented on src/Raven.Client/Documents/Subscriptions/DocumentSubscriptions.cs:
```diff
@@ -481,17 +481,43 @@ public List<SubscriptionState> GetSubscriptions(int start, int take, string data
         /// <param name="id"></param>
         /// <param name="database"></param>
-        public async Task DropConnectionAsync(string name, string database = null, CancellationToken token = default)
+        public async Task DropConnectionAsync(string name, long connectionId, string database = null, CancellationToken token = default)
```
breaks source-level compatibility

lastav5 replied:
Fixed

ppekrol commented on src/Raven.Server/Documents/Handlers/SubscriptionsHandler.cs:
```diff
@@ -381,18 +399,30 @@ public Task DropSubscriptionConnection()
                     .SubscriptionStorage
-                    .GetRunningSubscription(context, subscriptionId, subscriptionName, false);
-
+                    .GetRunningSubscription(context, subscriptionId, subscriptionName, false); //TODO stav: should consider subscription running if no connections exist at the time?
```
isn't that the same as in pre 5.3 impl when the connection was `null` ?

ppekrol replied:
does it matter for this code?

ml054 committed:
RavenDB-17360 Concurrent subscriptions Studio work

lastav5 committed:
RavenDB-15919 - Fixed dropping single connection through studio

ml054 committed:
RavenDB-17360 Concurrent subscriptions Studio work

karmeli87 committed:
RavenDB-15919 more abstraction and handle edge cases

ml054 committed:
RavenDB-17360 Concurrent subscriptions Studio work - show subscription mode

ayende commented on src/Raven.Client/Documents/Subscriptions/DocumentSubscriptions.cs:
```diff
@@ -481,17 +481,43 @@ public List<SubscriptionState> GetSubscriptions(int start, int take, string data
         /// <param name="id"></param>
         /// <param name="database"></param>
-        public async Task DropConnectionAsync(string name, string database = null, CancellationToken token = default)
+        public async Task DropConnectionAsync(string name, long connectionId, string database = null, CancellationToken token = default)
```
This is a breaking change, no?

lastav5 replied:
Fixed

ayende commented on src/Raven.Client/Documents/Subscriptions/SubscriptionWorker.cs:
```diff
@@ -112,7 +112,11 @@ public async ValueTask DisposeAsync(bool waitForSubscriptionTask)
+                        if (await _subscriptionTask.WaitWithTimeout(TimeSpan.FromSeconds(60)).ConfigureAwait(false) == false)
+                        {
+                            if (_logger.IsInfoEnabled)
+                                _logger.Info($"Subscription worker for '{SubscriptionName}' wasn't done after 60 seconds", new TimeoutException());
```
```suggestion
                                _logger.Info($"Subscription worker for '{SubscriptionName}' wasn't done after 60 seconds, cannot hold subscription disposal any longer.");
```

lastav5 replied:
Done

ayende commented on src/Raven.Server/Documents/Handlers/SubscriptionsHandler.cs:
```diff
@@ -381,18 +399,30 @@ public Task DropSubscriptionConnection()
                     .SubscriptionStorage
-                    .GetRunningSubscription(context, subscriptionId, subscriptionName, false);
-
+                    .GetRunningSubscription(context, subscriptionId, subscriptionName, false); //TODO stav: should consider subscription running if no connections exist at the time?
```
A subscription is running only if it has active connections

lastav5 replied:
Maybe a better question is.. Since we are in a subscription drop endpoint, don't we want to drop the subscription regardless if it's active or not?

ayende replied:
What does it mean, to have a running subscription without connections? what is inactive running subscription?

lastav5 replied:
That the connections list is empty but the SubscriptionConnectionsState instance still exists in the SubscriptionStorage list. So the subscription exists but at this point in time no connection is connected

ayende commented on src/Raven.Server/Documents/Handlers/SubscriptionsHandler.cs:
```diff
@@ -381,18 +399,30 @@ public Task DropSubscriptionConnection()
+                    if (connectionId.HasValue)
+                    {
+                        result = Database.SubscriptionStorage.DropSingleSubscriptionConnection(subscription.SubscriptionId, connectionId,
+                        new SubscriptionClosedException($"Connection with Id {connectionId} dropped by API request"));
```
Let's add some details here, who was the certificate that requested and where it came from

lastav5 replied:
Done

ayende commented on src/Raven.Server/Documents/Handlers/SubscriptionsHandler.cs:
```diff
@@ -381,18 +399,30 @@ public Task DropSubscriptionConnection()
+                    else 
+                    {
+                       result = Database.SubscriptionStorage.DropSubscriptionConnections(subscription.SubscriptionId,
+                           new SubscriptionClosedException("Dropped by API request"));
```
Let's add some details here, who was the certificate that requested and where it came from

lastav5 replied:
Done

ayende commented on src/Raven.Server/Documents/Subscriptions/LiveSubscriptionPerformanceCollector.cs:
```diff
@@ -61,57 +61,53 @@ protected IEnumerable<SubscriptionTaskPerformanceStats> PrepareInitialPerformanc
                             connectionAggregators.Add(currentConnection.GetPerformanceStats());
+                            
+                            // add connection stats to results 
+                            var subscriptionItem = results.Find(x => x.TaskId == kvp.Value.Handler.SubscriptionId);
```
Let's avoid `O(N^2)` op here. Make `results` into a dictionary, instead

ayende commented on src/Raven.Server/Documents/Subscriptions/SubscriptionProcessor.cs:
```diff
@@ -0,0 +1,152 @@
+
+        public override Task<long> RecordBatch(string lastChangeVectorSentInThisBatch)
+        {
+            // we skipped all documents, but we still need to update the change vector
```
?? What do you mean here?

karmeli87 replied:
the comment no longer relevant, but we might still record a batch with 0 docs, just to move forward the change vector

ayende replied:
I'm not seeing where that is done, though?

karmeli87 replied:
The `BatchItems` list is empty, but we still send the command.

ayende commented on src/Raven.Server/Documents/Subscriptions/SubscriptionProcessor.cs:
```diff
@@ -0,0 +1,152 @@
+                    }
+                }
+            }
+
```
`            ItemsToRemoveFromResend.Clear();` here - so we won't keep the string references for long

karmeli87 replied:
line 65 is clearing `ItemsToRemoveFromResend` I guess we can also clear the `BatchItems` after the actual ack

ayende replied:
You are clearing _before_, so the strings will be held until the _next_ batch

ayende commented on src/Raven.Server/Documents/Subscriptions/SubscriptionProcessor.cs:
```diff
@@ -0,0 +1,152 @@
+        }
+    }
+
+    public class DocumentsSubscriptionProcessor : SubscriptionProcessor<DocumentRecord>
```
Can this be used concurrently from multiple connections?

karmeli87 replied:
No, this is scoped to be per connection.

ayende commented on src/Raven.Server/Documents/Subscriptions/SubscriptionStorage.cs:
```diff
@@ -95,31 +96,53 @@ public async Task<long> PutSubscription(SubscriptionCreationOptions options, str
+        {
+            var command = new RecordBatchSubscriptionDocumentsCommand(_db.Name, subscriptionId, subscriptionName, list, previouslyRecordedChangeVector, lastRecordedChangeVector, _serverStore.NodeTag, _serverStore.LicenseManager.HasHighlyAvailableTasks(), RaftIdGenerator.NewId());
+            var (etag, _) = await _serverStore.SendToLeaderAsync(command);
+            await _db.RachisLogIndexNotifications.WaitForIndexNotification(etag, _serverStore.Engine.OperationTimeout);
```
Do we need this? I don't believe that we are actually reading anything here?

ayende commented on src/Raven.Server/Documents/Subscriptions/SubscriptionStorage.cs:
```diff
@@ -95,31 +96,53 @@ public async Task<long> PutSubscription(SubscriptionCreationOptions options, str
+            var command = new RecordBatchSubscriptionDocumentsCommand(_db.Name, subscriptionId, subscriptionName, list, previouslyRecordedChangeVector, lastRecordedChangeVector, _serverStore.NodeTag, _serverStore.LicenseManager.HasHighlyAvailableTasks(), RaftIdGenerator.NewId());
+            command.Deleted = deleted;
+            var (etag, _) = await _serverStore.SendToLeaderAsync(command);
+            await _db.RachisLogIndexNotifications.WaitForIndexNotification(etag, _serverStore.Engine.OperationTimeout);
```
Do we need this? I don't believe that we are actually reading anything here?

ayende commented on src/Raven.Server/Documents/Subscriptions/SubscriptionStorage.cs:
```diff
@@ -268,30 +291,35 @@ public async Task DeleteSubscription(string name, string raftRequestId)
 
             if (_logger.IsInfoEnabled)
-                _logger.Info($"Subscription with id '{subscriptionId}' and name '{subscriptionConnectionState.SubscriptionName}' connection was dropped. Reason: {ex.Message}");
+                _logger.Info($"Subscription with id '{subscriptionId}' and name '{subscriptionConnectionsState.SubscriptionName}' connections were dropped. Reason: {ex.Message}");
```
Put the _whole_ exception in. We *never* want to use `ex.Message`

lastav5 replied:
Done

ayende commented on src/Raven.Server/Documents/Subscriptions/SubscriptionStorage.cs:
```diff
@@ -268,30 +291,35 @@ public async Task DeleteSubscription(string name, string raftRequestId)
+            subscriptionConnectionsState.DropSingleConnection(connectionToDrop);
+            if (_logger.IsInfoEnabled)
+                _logger.Info(
+                    $"Connection with id {connectionId} in subscription with id '{subscriptionId}' and name '{subscriptionConnectionsState.SubscriptionName}' was dropped. Reason: {ex.Message}");
```
Same issue, --`ex.Message`--

lastav5 replied:
Done

ayende commented on src/Raven.Server/Documents/Subscriptions/SubscriptionStorage.cs:
```diff
@@ -335,14 +363,11 @@ public string GetSubscriptionNameById(TransactionOperationContext serverStoreCon
-                if (subscriptionStateConnection == null)
+                var subscriptionConnectionsState = kvp.Value;
+                
+                if (subscriptionConnectionsState.IsSubscriptionActive() == false)
```
When would that be false?

ayende commented on src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs:
```diff
@@ -315,88 +322,89 @@ private async Task FlushBufferToNetwork()
-                            connection._lastConnectionStats = new SubscriptionConnectionStatsAggregator(_connectionStatsId, null);
-                            connection._connectionScope = connection._lastConnectionStats.CreateScope();
+                        throw new SubscriptionClosedException(
+                            $"Cannot open new subscription connection, max amount of concurrent connections reached ({TcpConnection.DocumentDatabase.Configuration.Subscriptions.MaxNumberOfConcurrentConnections})");
```
Add the configuration entry name for setting this

ayende commented on src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs:
```diff
@@ -315,88 +322,89 @@ private async Task FlushBufferToNetwork()
+                    }
+                    catch (SubscriptionInvalidStateException)
+                    {
+                        _pendingConnectionScope.Dispose();
```
Why it this disposed only in this case? What about all the error potential scenarios that can throw here?

ayende commented on src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs:
```diff
@@ -315,88 +322,89 @@ private async Task FlushBufferToNetwork()
+                    if (e is SubscriptionInUseException)
+                        _connectionScope.RecordException(SubscriptionError.ConnectionRejected, e.Message);
+                    else
+                        _connectionScope.RecordException(SubscriptionError.Error, e.Message);
```
full error here

ayende commented on src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs:
```diff
@@ -315,88 +322,89 @@ private async Task FlushBufferToNetwork()
+                catch (Exception e)
+                {
+                    if (e is SubscriptionInUseException)
+                        _connectionScope.RecordException(SubscriptionError.ConnectionRejected, e.Message);
```
Full error here

ayende commented on src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs:
```diff
@@ -315,88 +322,89 @@ private async Task FlushBufferToNetwork()
+                }
+                finally
+                {
+                    _subscriptionConnectionsState?.DropSingleConnection(this); //TODO stav: _subscriptionConnectionsState might be disposed
```
You need a `try/catch` here, no? Or making sure that `DropSingleConnection` can never throw (preferred)

ayende commented on src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs:
```diff
@@ -670,25 +657,18 @@ private async Task ProcessSubscriptionAsync()
                                     }
-                                    AddToStatusDescription($"Acknowledging docs processing progress without sending any documents to client. CV: {_lastChangeVector}");
-
-                                    await TcpConnection.DocumentDatabase.SubscriptionStorage.AcknowledgeBatchProcessed(SubscriptionId,
```
what replaces this?

karmeli87 replied:
the processor `AcknowledgeBatch`

ayende replied:
But what will call it, it isn't from here

karmeli87 replied:
Okay, previously we were updating the change vector via the `AcknowledgeBatchProcessed` now we do that via the `RecordBatch` command, so basically the replacement for that is sending an empty `RecordBatch` command

ayende commented on src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs:
```diff
@@ -697,17 +677,17 @@ private async Task ProcessSubscriptionAsync()
-                                        if (globalEtag > _startEtag)
-                                            continue;
+                                        // TODO : would collection @all_docs work here as expected?
+                                        var globalEtag =_processor.GetLastItemEtag(docsContext, Subscription.Collection);
```
Not unless you handled that explicitly. Don't we have tests for `from @all_docs` ?

karmeli87 replied:
It seems we don't have tests to check this code path with `@all_docs`, but I've fix it anyway.

ayende commented on src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs:
```diff
@@ -720,8 +700,14 @@ private async Task ProcessSubscriptionAsync()
+                        }
                         catch (Exception e)
                         {
+                            _subscriptionConnectionsState.DropSingleConnection(this); //TODO stav: could remove this and only rely on outer 'finally' clause to drop the connection
```
Very likely, yes. Why have both?

lastav5 replied:
Fixed

ayende commented on src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs:
```diff
@@ -820,6 +803,28 @@ private void AssertCloseWhenNoDocsLeft()
             return (replyFromClientTask, subscriptionChangeVectorBeforeCurrentBatch);
         }
 
+        private bool ShouldAddToResendTable(DocumentsStorage.DocumentOrTombstone item, string currentChangeVector)
```
Isn't this duplicate code, we have that in the documents processor, no?

karmeli87 replied:
not quiet... we want to add here items that we might have skipped, because they were updated.
While the other function is to fetch items from the resend that aren't processed by an active connection. (e.g. due to a failure)

ayende commented on src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs:
```diff
@@ -829,195 +834,204 @@ private void AssertCloseWhenNoDocsLeft()
+                                {
+                                    if (result.Doc.Data.Modifications != null)
+                                    {
+                                        result.Doc.Data = docsContext.ReadObject(result.Doc.Data, "subsDocAfterModifications");
```
`subsDocAfterModifications` ??

karmeli87 replied:
it is a debug string

ayende commented on src/Raven.Server/ServerWide/Commands/Subscriptions/AcknowledgeSubscriptionBatchCommand.cs:
```diff
@@ -60,16 +72,84 @@ protected override BlittableJsonReaderObject GetUpdatedValue(long index, RawData
+                throw new RachisApplyException($"'{nameof(DatabaseName)}' is missing in '{nameof(AcknowledgeSubscriptionBatchCommand)}'.");
+            }
+
+            if (BatchId == default)
```
Are you comparing to null or 0 here?

karmeli87 replied:
`null`, will be more explicit

ayende commented on src/Raven.Server/ServerWide/Commands/Subscriptions/AcknowledgeSubscriptionBatchCommand.cs:
```diff
@@ -60,16 +72,84 @@ protected override BlittableJsonReaderObject GetUpdatedValue(long index, RawData
+
+                    tvb.Add(keySlice);
+                    tvb.Add(changeVectorSlice);
+                    tvb.Add(Bits.SwapBytes(index)); // batch id
```
Maybe a note that this is a *non existent* batch id

ayende commented on src/Raven.Server/ServerWide/Commands/Subscriptions/RecordBatchSubscriptionDocumentsCommand.cs:
```diff
@@ -0,0 +1,234 @@
+                var existingValue = new BlittableJsonReaderObject(ptr, size, context);
+
+                if (existingValue == null)
+                    throw new SubscriptionDoesNotExistException($"Subscription with name '{subscriptionName}' does not exist");
```
In what database?

lastav5 replied:
Fixed

ayende commented on src/Raven.Server/ServerWide/Commands/Subscriptions/RecordBatchSubscriptionDocumentsCommand.cs:
```diff
@@ -0,0 +1,234 @@
+                if (subscriptionState.ChangeVectorForNextBatchStartingPoint != PreviouslyRecordedChangeVector)
+                {
+                    throw new SubscriptionChangeVectorUpdateConcurrencyException($"Can't record subscription with name '{subscriptionName}' due to inconsistency in change vector progress. Probably there was an admin intervention that changed the change vector value. Stored value: {subscriptionState.ChangeVectorForNextBatchStartingPoint}, received value: {PreviouslyRecordedChangeVector}");
+                    //TODO stav: might need to delete entire table if admin changed cv
```
That is a really good point. the question here is what did they change it _to_ ? 
What is the scenarios we have to deal with here? What do we expect to happen?

karmeli87 replied:
I think we just need to delete the entire table.
This basically tells us - start the subscription from scratch at the given starting point, no?

ayende replied:
You cannot delete the actual table, since this is a global shared table.
But agree, we should probably clear all the items for that subscription in this case. However, not sure if the user is actually interested in that scenario.
Or at least, skip all those that are _after_ the current value.

karmeli87 replied:
I can delete by prefix... `DatabaseName/SubId/`

ayende replied:
I have no issue with this if the user goes to the beginning of time.
But what about when they just change the CV? I guess that we should trust the user, and just discard everything else?

Would we need to warn the user if there are items in the resend queue?

karmeli87 replied:
Hmm.. we can clear everything above the new change vector then?
But I'm not sure what the user expect at this point

ayende replied:
We can easily do that if the new change vector precede the resend items. 
However, the real question is indeed what do we expect the *user* to think should happen.

The more I think about it, the more I tend toward saying, the user wanted to *reset* the position of the subscription, so anything pending is going to be cleared, which is the simplest to implement and to explain.

ayende replied:
Also remember that the resend queue is an *implementation detail*, we aren't showing that to the user in any way

karmeli87 replied:
I wonder if we want to have a dedicated EP/command for changing the starting point change vector by the admin.
My concern here is that due to a bug we might delete those entries..

ayende commented on src/Raven.Server/ServerWide/Commands/Subscriptions/RecordBatchSubscriptionDocumentsCommand.cs:
```diff
@@ -0,0 +1,234 @@
+
+        protected override BlittableJsonReaderObject GetUpdatedValue(long index, RawDatabaseRecord record, JsonOperationContext context, BlittableJsonReaderObject existingValue)
+        {
+            throw new System.NotImplementedException();
```
??

karmeli87 replied:
This used in the base `Execute`, but we override it here and not using those methods.

ayende commented on src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionsState.cs:
```diff
@@ -0,0 +1,590 @@
+            PreviouslyRecordedChangeVector = LastChangeVectorSent;
+        }
+
+        public static SubscriptionConnectionsState CreateDummyState(DocumentsStorage storage, SubscriptionState state) => new (storage, state);
```
who is calling this, and why?

karmeli87 replied:
This is used for subscription test endpoint

ayende commented on src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionsState.cs:
```diff
@@ -0,0 +1,590 @@
+                using (GetDatabaseAndSubscriptionKeyPrefix(context, DocumentDatabase.Name, SubscriptionId, SubscriptionType.Document, out var prefix))
+                using (Slice.External(context.Allocator, prefix, out var prefixSlice))
+                {
+                    return subscriptionState.SeekByPrimaryKeyPrefix(prefixSlice, Slices.Empty, 0).Count();
```
This is an `O(N)` operation, which is not something that we really want. 
Why is this needed?

karmeli87 replied:
This used only for testing

ayende commented on src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionsState.cs:
```diff
@@ -0,0 +1,590 @@
+                SubscriptionId,
+                SubscriptionName,
+                connection.LastSentChangeVectorInThisConnection ?? nameof(Client.Constants.Documents.SubscriptionChangeVectorSpecialStates.DoNotChange),
+                "",
```
What is this?

karmeli87 replied:
we no longer need to send this param.
refactored this part.

ayende commented on src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionsState.cs:
```diff
@@ -0,0 +1,590 @@
+
+        public bool IsSubscriptionActive()
+        {
+            return _connections.Count != 0;
```
when *won't* it be the case?

lastav5 replied:
There can be cases when all connections disconnected but the SubscriptionConnectionsState still exists in the the SubscriptionStorage list

ayende commented on src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionsState.cs:
```diff
@@ -0,0 +1,590 @@
+            return _connections.Count != 0;
+        }
+
+        internal Document GetRevision(string changeVector,DocumentsOperationContext context)
```
Reverse args

karmeli87 replied:
removed an unused method.

ayende commented on src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionsState.cs:
```diff
@@ -0,0 +1,590 @@
+
+        private void RegisterSingleConnection(SubscriptionConnection incomingConnection)
+        {
+            lock (this)
```
Why is this needed? Already we already locked?

karmeli87 committed:
RavenDB-17360 move some files around and address PR comments

karmeli87 committed:
RavenDB-17360 upgrade server schema

karmeli87 committed:
RavenDB-17360 fix test subscription processor

karmeli87 committed:
RavenDB-17360 fix backward comp.

lastav5 committed:
RavenDB-15919 - Cancel connection on drop. Address PR comments. Client API connection drop fix

lastav5 committed:
RavenDB-15919 - Addressed PR comments. Fixes

karmeli87 committed:
RavenDB-15919 adjust subscription behaviour

- Putting existing subscription will delete the resend table
- Force reconnect if we modified the subscription
- Upon connecting first update client connection time and then send a noop
- handle dropping the connection properly
- create EP to expose the resend table

karmeli87 committed:
RavenDB-15919 check for query change before LastClientConnectionTime

karmeli87 committed:
RavenDB-15919 fix EP

karmeli87 committed:
RavenDB-15919 change connectionId to workerId

ppekrol commented on src/Raven.Client/Documents/Commands/DropSubscriptionConnectionCommand.cs:
```diff
@@ -9,11 +9,17 @@ namespace Raven.Client.Documents.Commands
         {
             _name = name;
         }
+        public DropSubscriptionConnectionCommand(string name, string workerId)
```
space + internal?

ppekrol replied:
call the other ctor: `this(name)`

ppekrol replied:
maybe check for null or whitespace on name as well?

ppekrol commented on src/Raven.Client/Documents/Subscriptions/SubscriptionWorkerOptions.cs:
```diff
@@ -110,6 +110,7 @@ public SubscriptionWorkerOptions(string subscriptionName) : this()
                 throw new ArgumentException("Value cannot be null or empty.", nameof(subscriptionName));
 
             SubscriptionName = subscriptionName;
+            WorkerId = Guid.NewGuid().ToString();
```
it should be move to private ctor?

karmeli87 replied:
we don't want to generate `WorkerId` upon de-serializing.

karmeli87 committed:
RavenDB-15919 address PR comments

karmeli87 committed:
RavenDB-15919 add license check

ppekrol commented on src/Raven.Server/Documents/Handlers/SubscriptionsHandler.cs:
```diff
@@ -213,6 +223,31 @@ await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBody
+
+                var items = SubscriptionStorage.GetResendItems(context, Database.Name, subscriptionState.SubscriptionId);
+
+                writer.WriteArray("Items", items.Select(i => i.ToJson()), context);
```
this does not generate proper Json, we are missing start and end object here. Also most of the times this is called 'Results' ?
